Fix #3304: Add check to ensure KDoc presence for non-private members

Sparsh1212 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #3304: Add check to ensure KDoc presence for non-private members

Script for ensuring that KDocs are present on all non-private:
- Classes
- Functions
- Values/fields
- Explicit constructors
- Companion objects
- Nested classes
- Enums
- Annotations
- Interfaces

Note: If any of the above member has an override modifier present, then it is automatically
exempted for the KDoc check.

Usage:
`bazel run //scripts:kdoc_check -- <path_to_directory_root>`

Arguments:
- path_to_directory_root: directory path to the root of the Oppia Android repository.

Example:
`bazel run //scripts:kdoc_check -- $(pwd)`

For testing the script, automated tests have been added.
To execute the tests, use:

`bazel test //scripts/src/javatests/org/oppia/android/scripts/kdoc:KDocCheckTest`

**Note:** We are generating the test assets dynamically at the time of executing them. The test assets are automatically deleted when the test finishes.

SS when the CI check fails:

![image](https://user-images.githubusercontent.com/55937724/126044224-5cc1c960-d8e5-4bbb-a3da-e05b92bef934.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sparsh1212 committed:
Fix #3290: Add support for generic regex pattern matching

Sparsh1212 committed:
Apply review suggestions for proto location

Sparsh1212 committed:
Delete main directory

Sparsh1212 committed:
Implment script logic and tests

Sparsh1212 committed:
CI setup part 1

Sparsh1212 committed:
syntax fix in yaml file

Sparsh1212 committed:
import nits in dummy file

Sparsh1212 committed:
Use regex patterns from script in script test

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Improve naming of script_assets variables

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Revamp testing approach

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Fix #3291: Add check for XML syntax correctness

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Introduce XML syntax check in static checks workflow

Sparsh1212 committed:
Fix #3292: Add check for test files presence for prod files

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Make nit suggestions

Sparsh1212 committed:
Fix #3300: Add check for accessibility labels for activities

Sparsh1212 committed:
Add CI check for label presence for activities

Sparsh1212 committed:
Add ScriptConstants file to exemptions test list

Sparsh1212 committed:
merge test-file-check

Sparsh1212 committed:
Test if: always()

Sparsh1212 committed:
Revert "Test if: always()"

This reverts commit b72e419b

Sparsh1212 committed:
Add if: always()

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Add if: always()

Sparsh1212 committed:
Merge branch 'test-file-check' into accesibilty-label-check

Sparsh1212 committed:
Add if: always()

Sparsh1212 committed:
Apply review suggestions on PR

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make review suggestions from regex pattern checks

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Implement review suggestions based on #3340

Sparsh1212 committed:
Refactor PR as per feedback recieved

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement feedback suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nits

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Nit fixes

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
merge test-file-check

Sparsh1212 committed:
Implement feedback suggestions as per #3340

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Merge branch 'test-file-check' into accesibilty-label-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Merge branch 'test-file-check' into accesibilty-label-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Review suggestions part 1

Sparsh1212 committed:
Implement review suggestions part 2

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
resolve merge conflict

Sparsh1212 committed:
update static_checks

Sparsh1212 committed:
add test to repository file

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
Do review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
Implement review suggestions

Sparsh1212 committed:
Add Ben as a codeowner for the ScriptExemptions file

Sparsh1212 committed:
bazel files nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
Introduce a library for the regex assets

Sparsh1212 committed:
Make directory structure same as that of #3374

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Make directory structure similar to #3374

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
merge xml-syntax-checks

Sparsh1212 committed:
add new files to test file script exemptions and nit fixes

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
diable ktlint max-line-length

Sparsh1212 committed:
disable ktlint-max-line

Sparsh1212 committed:
disable ktlint max-length

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
remove testonly attribute from tests

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add new files to script exemptions

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
merge regex-checks

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Apply review suggestions

Sparsh1212 committed:
ktlint fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 committed:
Apply nit fixes

Sparsh1212 committed:
Remove script constants

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
add tests for XML syntax error handler

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
add todo

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add todo

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'develop' into regex-checks

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
add test

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
add testOnly

Sparsh1212 committed:
Merge branch 'regex-checks' into xml-syntax-check

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
fix textproto file

Sparsh1212 committed:
merge-regex-checks

Sparsh1212 committed:
Merge branch 'xml-syntax-check' into test-file-check

Sparsh1212 committed:
merge xml-syntax-check

Sparsh1212 committed:
add exemptions to textproto

Sparsh1212 committed:
sort failures lexicographically

Sparsh1212 committed:
Merge branch 'develop' into test-file-check

Sparsh1212 committed:
add new files to exemption list

Sparsh1212 committed:
merge test-file-check

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
add kdocs

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Trigger CI failure exclusively

Sparsh1212 committed:
Revert CI check failure

Sparsh1212 committed:
Fix #3340: Create a script to ensure KDoc presence

Sparsh1212 committed:
Add initial tests

Sparsh1212 committed:
Add more tests and add documentation

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge develop

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
merge accessibility label check

Sparsh1212 committed:
nit fix

Sparsh1212 committed:
Add more files to kdoc exemptions

Sparsh1212 commented:
@BenHenning PTAL.

BenHenning commented:
Will need to take a look tomorrow, I think. Apologies for the delay.

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

BenHenning commented:
Apologies, I think I might need to delay reviewing this one more day.

BenHenning commented on .github/workflows/static_checks.yml:
```diff
@@ -123,3 +123,8 @@ jobs:
+      - name: KDoc Check
+        if: always()
+        run: |
+          bazel run //scripts:kdoc_check -- $(pwd)
```
Maybe instead: 'KDoc validity check' would be clearer? The current name is a bit too non-specific, and we do want to eventually extend this script to include other checks.

Ditto for the step name above & class/library names elsewhere in the PR.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+"""
+Libraries corresponding to kdoc presence check, that ensures all public members (except overrides)
```
This & the test counterpart should be more generic: this package contains scripts corresponding to KDocs in the codebase. Presence is one such example, and there may be more in the future.

That being said, I actually suggest making this package even more generic. Could we call it 'docs'? We probably will only ever have one script for verifying KDocs, but I could see the need for validating other forms of documentation & comments in the future.

Sparsh1212 replied:
Done.

BenHenning commented on third_party/versions.bzl:
```diff
@@ -90,6 +90,7 @@ MAVEN_TEST_DEPENDENCY_VERSIONS = {
     "com.squareup.okhttp3:mockwebserver": "4.1.0",
     "com.squareup.retrofit2:retrofit-mock": "2.5.0",
     "junit:junit": "4.12",
+    "org.jetbrains.kotlin:kotlin-compiler-embeddable": "1.5.0",
```
@prayutsu will we need to also update the license textproto file, or will you be doing that in a later PR (given that there's no CI check currently to ensure it stays up-to-date?).

prayutsu replied:
As this is a test dependency, we'll not need to update the textproto file.

BenHenning replied:
Ah good point. Thanks!

BenHenning commented on scripts/script_assets.bzl:
```diff
@@ -112,3 +112,27 @@ def generate_accessibility_label_assets_list_from_text_protos(
+        name,
+        kdoc_exemptions_name):
+    """
+    Converts multiple lists of text proto assets to binary.
```
Is this docstring correct?

Sparsh1212 replied:
Corrected now.

BenHenning commented on scripts/script_assets.bzl:
```diff
@@ -112,3 +112,27 @@ def generate_accessibility_label_assets_list_from_text_protos(
+
+    Args:
+        name: str. The name of this generation instance. This will be a prefix for derived targets.
+        kdoc_exemptions_name: list of str. The list of kdoc exemptions file names.
```
It's just one file, right?

Sparsh1212 replied:
Corrected now.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+}
+
+/**
+ * Genrates a [KtFile] from a Kotlin file.
```
Typo: Generates

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+ *
+ * @return the string equivalent
+ */
+private fun File.asString(): String {
+  val fileContents = StringBuilder(this.length().toInt())
+  Scanner(this).use { scanner ->
+    while (scanner.hasNextLine()) {
+      fileContents.append(scanner.nextLine() + System.lineSeparator())
+    }
+    return fileContents.toString()
+  }
+}
```
Could we use [``File.readText()``](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/java.io.-file/read-text.html) instead?

Sparsh1212 replied:
Actually, if we use `File.readText()`, then it gives this error while parsing one of the files because of the line separator, hence I had to go by this approach.

```
Exception in thread "main" java.lang.AssertionError: Wrong line separators: '....gae.model\r\n\r\nimport...' at offset 49
        at org.jetbrains.kotlin.com.intellij.openapi.util.text.StringUtil.assertValidSeparators(StringUtil.java:2450)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.assertValidSeparators(DocumentImpl.java:690)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.<init>(DocumentImpl.java:119)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.<init>(DocumentImpl.java:114)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.<init>(DocumentImpl.java:105)
        at org.jetbrains.kotlin.com.intellij.core.CoreApplicationEnvironment.lambda$new$1(CoreApplicationEnvironment.java:94)
        at org.jetbrains.kotlin.com.intellij.mock.MockFileDocumentManagerImpl.getDocument(MockFileDocumentManagerImpl.java:42)
        at org.jetbrains.kotlin.com.intellij.psi.AbstractFileViewProvider.getDocument(AbstractFileViewProvider.java:180)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.retrieveLineNumberForElement(KDocValidityCheck.kt:260)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.checkIfKDocMissing(KDocValidityCheck.kt:208)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.elementIterator(KDocValidityCheck.kt:169)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.hasKdocFailure(KDocValidityCheck.kt:130)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.main(KDocValidityCheck.kt:86)
```

BenHenning replied:
Ah. You could maybe do something like:

```kotlin
file.readLines().joinToString(separateor = System.lineSeparator())
```

Would that work?

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+private fun hasKdocFailure(file: File, kDocNotRequiredAnnotationEntryList: List<String>): Boolean {
+  val ktFile = createKtFile(file.asString(), file.name)
+  return ktFile.children.fold(initial = false) { isFailing, elem ->
+    val childFailingKdoc = if (elem is KtElement) elementIterator(
```
This if-check should have curly braces. We only omit curly braces on statements if they fit entirely on one line.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+  file: File,
+  kDocNotRequiredAnnotaionEntryList: List<String>
+): Boolean {
+  if (elem is KtClass) {
```
I suggest converting this if-else-if block into a ``when`` block.

Sparsh1212 replied:
Done.

Sparsh1212 committed:
Merge branch 'develop' into accesibilty-label-check

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+ *     on an element, the element does not needs to be checked for a KDoc.
+ * @return whether the element is missing a KDoc
+ */
+private fun elementIterator(
```
I'm not sure how to understand what this method is meant to do from the name alone. Could you rename it to be something more specific as to what it's actually doing? Generally, method names are verb phrases since they describe an action that's taking place.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+ *     on an element, the element does not needs to be checked for a KDoc.
+ * @return whether the element is missing a KDoc
+ */
+private fun checkIfKDocMissing(
```
```suggestion
private fun checkIfKDocIsMissing(
```

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+      isMissingKdoc || childMissingKdoc
+    }
+    return memberMissingKdoc
+  } else if (elem is KtNamedFunction ||
+    elem is KtVariableDeclaration ||
+    elem is KtSecondaryConstructor
```
Can this fit on one line?

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+      isMissingKdoc || childMissingKdoc
+    }
+    return classKdocMissing || memberMissingKdoc
+  } else if (elem is KtObjectDeclaration && elem.isCompanion()) {
```
Is the companion check needed? I'm guessing regular object declarations should also have KDocs.

Sparsh1212 replied:
As per the discussion, we had, its now updated.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+}
+
+/**
+ * Returns whether an element is required to be checked for a KDoc presence
```
```suggestion
 * Returns whether an element is required to be checked for a KDoc presence.
```

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+ */
+private fun retrieveLineNumberForElement(statement: KtElement, markEndOffset: Boolean): Int? {
+  val file = statement.containingFile
+  if (file is KtFile && file.doNotAnalyze != null) {
```
Why is the doNotAnalyze check needed?

Sparsh1212 replied:
Apologies! this was redundant. Removed.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+}
+
+/**
+ * Loads the KDoc exemptions list to proto.
```
From a proto file?

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+ * @return proto class from the parsed textproto file
+ */
+private fun loadKdocExemptionsProto(kdocExemptiontextProto: String): KDocExemptions {
+  val protoBinaryFile = File("$kdocExemptiontextProto.pb")
```
Typo: TextProto

Ditto elsewhere.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/kdoc_exemptions.proto:
```diff
@@ -0,0 +1,22 @@
+  // List of all the files which should be exempted for the KDoc check.
+  // For exempting any file, provide its relative path to root in
+  // 'script/assets/kdoc_exemptions.textproto'.
+  // Also, note that the exemptions in the textproto file are maintained in lexicographical order.
```
```suggestion
  // Also, note that the exemptions in the text proto file are maintained in lexicographical order.
```

Ditto below.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/kdoc/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+"""
+Tests corresponding to kdoc presence check, that ensures all public members (except overrides) must
+have a corresponding KDoc.
```
As mentioned earlier, this docstring is a bit too specific to the exact check being implemented. Please generalize it.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/kdoc/KDocCheckTest.kt:
```diff
@@ -0,0 +1,487 @@
+import java.io.PrintStream
+
+/** Tests for [KDocCheck]. */
+class KDocCheckTest {
```
Other tests that I think are worth adding:
- A block comment instead of a KDoc (to ensure KDocs are specifically checked).
- Inject primary constructor (just to make sure nothing special is triggered in that situation)
- Object vals
- Constants
- Companion objects
- Nested interfaces
- Nested annotations
- All other special case annotations not already tested below (all of them ought to be checked)
- The special case of the exemptions list having a redundant file in it (see my comment above)

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt:
```diff
@@ -0,0 +1,284 @@
+  // Path of the repo to be analyzed.
+  val repoPath = "${args[0]}/"
+
+  val kdocExemptiontextProto = "scripts/assets/kdoc_exemptions"
```
One scenario I think we missed in past checks that is especially important for this one is outputting an error when the exemption list contains something no longer relevant. I think we actually want to make sure that, in general, exemption lists are always the complete picture on the status of the codebase.

Sparsh1212 replied:
Added.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212! Took a full pass.

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

vinitamurthi reviewed (approved):
LGTM for the codeowner files on the assumption that Ben's comments will be addressed

Sparsh1212 committed:
Add redundant exemptions check

Sparsh1212 committed:
Merge branch 'develop' into accesibilty-label-check

Sparsh1212 committed:
rectify static_checks

Sparsh1212 committed:
Remove duplicate proto libraries

Sparsh1212 committed:
merge accessibility-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes and add redundant exemptions check to script

Sparsh1212 commented:
@BenHenning PTAL.
- All comments addressed
- Added check for redundant exemptions
- Added tests for redundant exemptions
- Added new tests as suggested for KDocs

BenHenning commented on scripts/src/java/org/oppia/android/scripts/docs/BUILD.bazel:
```diff
@@ -1,18 +1,17 @@
 """
-Libraries corresponding to kdoc presence check, that ensures all public members (except overrides)
-must have a corresponding KDoc.
+Libraries corresponding docs validity check, that ensures the documentations are valid.
```
```suggestion
Libraries corresponding to documentation-related scripts.
```

Can be simpler. Ditto for the tests version.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/docs/KDocValidityCheck.kt:
```diff
@@ -1,4 +1,4 @@
-package org.oppia.android.scripts.kdoc
+package org.oppia.android.scripts.docs
```
File: ``KdocValidityCheck``

(note camel-casing/capitalization). Ditto for the proto message name.

We consider "kdoc" to be one word which means as a unit it's capitalized/lowercase rather than being partly.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/docs/KDocValidityCheck.kt:
```diff
@@ -72,24 +74,40 @@ fun main(vararg args: String) {
 
   // A list of all the files to be exempted for this check.
-  val kdocExemptionList = loadKdocExemptionsProto(kdocExemptiontextProto).getExemptedFilePathList()
+  val kdocExemptionList = loadKdocExemptionsProto(pathToProtoBinary)
+    .getExemptedFilePathList()
```
```suggestion
  val kdocExemptionList =
    loadKdocExemptionsProto(pathToProtoBinary).getExemptedFilePathList()
```

Wrapping should be at the higher syntactical element (in this case the ``=``).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/docs/KDocValidityCheck.kt:
```diff
@@ -122,17 +140,23 @@ private fun createKtFile(codeString: String, fileName: String): KtFile {
  * @param kDocNotRequiredAnnotationEntryList the list of annotation entries which when present
  *     on an element, the element does not needs to be checked for a KDoc.
- * @return whether the file has a KDoc presence failure
+ * @return Returns list of elements in the file which are missing KDoc
```
```suggestion
 * @return a list of elements in the file which are missing KDocs
```

Ditto elsewhere.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/docs/KDocValidityCheck.kt:
```diff
@@ -158,33 +182,52 @@ private fun File.asString(): String {
+          kDocNotRequiredAnnotationEntryList
+        )
+      }
+      return listOf(listOf(classKdocMissing), memberMissingKdoc).flatten()
```
Can be simplified I think to:

```suggestion
      return memberMissingKdoc + classKdocMissing
```

Ditto elsewhere.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/docs/KDocValidityCheck.kt:
```diff
@@ -122,17 +140,23 @@ private fun createKtFile(codeString: String, fileName: String): KtFile {
+        kDocNotRequiredAnnotationEntryList
+      )
+    } else {
+      listOf(null)
```
Why not return empty list, instead? Ditto below.

Prefer to avoid nullable unless needed.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/proto/script_exemptions.proto:
```diff
@@ -35,3 +35,19 @@ message AccessibilityLabelExemptions {
 }
+
+// Exemptions for the KDoc validity check. Exemptions indicate files for which we should
+// not check for the presence of a KDoc.
```
```suggestion
// not check for the presence of KDocs.
```

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -118,6 +123,119 @@ class KDocCheckTest {
+  }
+
+  @Test
+  fun testKDoc_fieldWithRuleAnnotationWithoutKDoc_checkShouldPass() {
```
```suggestion
  fun testKdoc_fieldWithRuleAnnotationWithoutKdoc_checkShouldPass() {
```

Ditto for other tests (per camel-casing guidelines).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -118,6 +123,119 @@ class KDocCheckTest {
+    val testContent =
+      """
+      @Provides
+      @ExplorationStorageDatabaseSize
```
What about a test where something just has a qualifier? That seems like a nice failure case to cover.

Sparsh1212 replied:
Added.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -154,6 +272,22 @@ class KDocCheckTest {
   }
 
+  @Test
+  fun testKDoc_injectPrimaryConstructor_noSpecialBehaviour_checkShouldPass() {
```
_withKdoc?

Ditto for other passing cases (to contrast with the 'withoutKdoc' tests). Similarly, I suggest separating out "_withoutKdoc" in those test names to be a bit clearer that it's a separate condition being implemented.

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -154,6 +272,22 @@ class KDocCheckTest {
+    val testContent =
+      """
+      /** Test Class. */
+      class SelectionFragmentModel @Inject constructor(){
```
```suggestion
      class SelectionFragmentModel @Inject constructor() {
```

Sparsh1212 replied:
Done.

BenHenning replied:
Was this addressed? The latest code version doesn't seem to include the space.

Sparsh1212 replied:
Apologies! Just missed this. Now updated.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -246,8 +383,84 @@ class KDocCheckTest {
+  fun testKDoc_blockCommentInsteadOfKDoc_checkShouldFail() {
+    val testContent =
+      """
+      // Test block comment.
```
This is a line comment, not a block comment (suggest updating the test name & adding a test for a block comment).

Sparsh1212 replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -463,25 +696,101 @@ class KDocCheckTest {
   }
 
+  @Test
+  fun testKDoc_redundantExemption_checkShouldFail() {
```
I suggest framing this as "fileWithAllExpectedKdocs_redundantExemption_..." and then add a test for the case where it's exempted and only has some of the kdocs--that shouldn't fail since the file hasn't been fully resolved yet.

Sparsh1212 replied:
Done.

BenHenning reviewed (changes requested):
Thanks @Sparsh1212. Had a few more comments--PTAL.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/docs/KDocValidityCheckTest.kt:
```diff
@@ -0,0 +1,796 @@
+  }
+
+  @Test
+  fun testKDoc_companionObjectMissingKDoc_checkShouldFail() {
```
I suggest making this test specific to just the companion object and add another test for a companion object with constituent values & functions to make it more clear the specific expected behavior for companion objects.

Sparsh1212 replied:
Done.

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
add activity to exemption

Sparsh1212 committed:
Merge branch 'accesibilty-label-check' into kdoc-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'accesibilty-label-check' into kdoc-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'accesibilty-label-check' into kdoc-check

Sparsh1212 committed:
nit fixes

Sparsh1212 committed:
Merge branch 'accesibilty-label-check' into kdoc-check

Sparsh1212 committed:
Merge branch 'develop' into accesibilty-label-check

Sparsh1212 committed:
update branch

Sparsh1212 committed:
Merge branch 'accesibilty-label-check' into kdoc-check

BenHenning reviewed (changes requested):
Thanks @Sparsh1212! I think this looks really great--the test name changes make them *much* clearer, I think.

I just noticed one comment didn't seem resolved, but I otherwise have nothing new to add.

Could you please fix the deltas on this PR & its title since #3352 was merged?

Sparsh1212 committed:
update branch

Sparsh1212 commented:
> Thanks @Sparsh1212! I think this looks really great--the test name changes make them _much_ clearer, I think.
> 
> I just noticed one comment didn't seem resolved, but I otherwise have nothing new to add.
> 
> Could you please fix the deltas on this PR & its title since #3352 was merged?

Thanks @BenHenning the PR is now updated and the one left a comment is also addressed. PTAL.
Also assigning to @anandwana001 for a pass.

Sparsh1212 committed:
Merge branch 'develop' into kdoc-check

anandwana001 commented on .github/workflows/static_checks.yml:
```diff
@@ -123,3 +123,8 @@ jobs:
         run: |
           bazel run //scripts:accessibility_label_check -- $(pwd) scripts/assets/accessibility_label_exemptions.pb app/src/main/AndroidManifest.xml
+
+      - name: KDoc validity Check
```
```suggestion
      - name: KDoc Validation Check
```

Sparsh1212 replied:
Done.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/docs/KdocValidityCheck.kt:
```diff
@@ -0,0 +1,347 @@
+import java.io.FileInputStream
+
+/**
+ * Script for ensuring that KDocs validity on all non-private:
```
```suggestion
 * Script for ensuring the KDocs validity on all non-private:
```

anandwana001 replied:
Also, make file name `KDocValidityCheck.kt`

Sparsh1212 replied:
Explained below. PTAL.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/docs/KdocValidityCheck.kt:
```diff
@@ -0,0 +1,347 @@
+ */
+fun main(vararg args: String) {
+  // Path of the repo to be analyzed.
+  val repoPath = "${args[0]}/"
```
use the same name given above in comment - `path_to_directory_root` in camel case

Sparsh1212 replied:
Actually, we are using the same name "repoPath" for all the scripts till now, so will it be better to keep this the same in order to maintain consistency across all the scripts?

anandwana001 commented on scripts/src/javatests/org/oppia/android/scripts/docs/KdocValidityCheckTest.kt:
```diff
@@ -0,0 +1,875 @@
+class KdocValidityCheckTest {
+  private val outContent: ByteArrayOutputStream = ByteArrayOutputStream()
+  private val originalOut: PrintStream = System.out
+  private val KDOC_CHECK_PASSED_OUTPUT_INDICATOR = "KDOC VALIDITY CHECK PASSED"
+  private val KDOC_CHECK_FAILED_OUTPUT_INDICATOR = "KDOC VALIDITY CHECK FAILED"
```
`private const val ` and make them global

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/javatests/org/oppia/android/scripts/docs/KdocValidityCheckTest.kt:
```diff
@@ -0,0 +1,875 @@
+        return error?.let(context::getString)
+      }
+      """.trimIndent()
+    val tempFile = tempFolder.newFile("testfiles/TempFile.kt")
+    tempFile.writeText(testContent)
```
you can create a helper function at bottom of this file and use it here for writing text file

Sparsh1212 replied:
Hmm, we just want to use `writeText` here in order to write the test file, so do we actually need to create a separate function for that? Also, this will make this script a bit inconsistent with all the other scripts, as there we are following the current approach.

anandwana001 commented on scripts/src/javatests/org/oppia/android/scripts/docs/KdocValidityCheckTest.kt:
```diff
@@ -0,0 +1,875 @@
+import java.io.PrintStream
+
+/** Tests for [KdocValidityCheck]. */
+class KdocValidityCheckTest {
```
```suggestion
class KDocValidityCheckTest {
```

Sparsh1212 replied:
Ah! this has been discussed previously, and it was decided to name it as "Kdoc", as "Kdoc" is a single entity in itself.
Ditto for the script name.

BenHenning replied:
For context, see also: https://developer.android.com/kotlin/style-guide#camel_case for why it should be 'Kdoc' and not 'KDoc' here.

anandwana001 reviewed (changes requested):
Thanks @Sparsh1212 
Took a quick pass on this. Left few suggestions. Also, please check on failing actions.

Sparsh1212 committed:
nit fixes

Sparsh1212 commented:
> Thanks @Sparsh1212
> Took a quick pass on this. Left few suggestions. Also, please check on failing actions.

Thanks, @anandwana001 followed up on all the suggestions. PTAL.

BenHenning reviewed (approved):
Thanks @Sparsh1212! LGTM.

anandwana001 reviewed (approved):
LGTM
